Improve mpi checks

drwells opened the pull request:
This PR adds a new exception `ExcMPI` that prints nice messages based on the MPI error code. It also checks the return codes of all MPI functions.

I have consistently used `AssertThrow`, not `Assert`, for these calls.

Fixes #3516.

bangerth commented on include/deal.II/base/exceptions.h:
```diff
@@ -1163,6 +1163,36 @@ namespace StandardExceptions
+ *
+ * This exception can be used when <code>deal.II</code> is compiled without
+ * MPI support: in that case it simply prints a message including the
+ * numerical value of the error code.
```
In what circumstances would we check error code then? Or are you simply saying that one can use the exception without mpi (even though that would be in a context in which it is clearly not appropriate)?

drwells replied:
What I did is probably unnecessary. We typedef `MPI_Op`, `MPI_Comm`, and `MPI_Datatype` if `DEAL_II_WITH_MPI` is false, so (following that idea) this exception is defined whether or not we have MPI (though it is not very useful when we do not have it).

I believe the correct answer to your question is 'never'.

bangerth commented on include/deal.II/base/mpi.templates.h:
```diff
@@ -95,21 +95,23 @@ namespace Utilities
+                              internal::mpi_type_id(values),
+                              mpi_op,
+                              mpi_communicator);
+            AssertThrow(ierr == MPI_SUCCESS, ExcMPI(ierr));
```
You could make all of these checks simpler if you had a macro that expands to this very line and only takes `ierr` add argument. I think we have such variations of Assert in exceptions.h for other repetitive cases.

drwells replied:
That's a good idea. I'll replace these.

tjhei replied:
Just to add another opinion here: I like your original, more verbose, version much better. Now anyone reading the code has to remember/lookup what this macro does.

bangerth commented on source/base/exceptions.cc:
```diff
@@ -336,6 +343,45 @@ void ExceptionBase::generate_message () const
+  // class.
+  int error_class = 0;
+  // there is nothing we can do if these functions fail (they just provide
+  // better input), so assume that the given error code is valid.
```
Output? 

I don't understand what these two lines comment on.  On the fact that the following mpi calls are not return value checked?

drwells replied:
The output is the message printed to the screen. I assumed that if we get this far then error code is a real MPI error code and the message printing functions will understand it.

Now that I look at things again I think it would be worthwhile to check these return codes and print a shorter message if one of them fails (which is unlikely to happen but the check is easy).

bangerth commented on source/base/exceptions.cc:
```diff
@@ -336,6 +343,45 @@ void ExceptionBase::generate_message () const
+      << std::endl
+      << "The numerical value of the original error code is "
+      << error_code
+      << "."
```
Yes, very nicely done with the translation of the error codes!

drwells replied:
Thanks! :)

drwells commented:
I rewrote things with a new macro `AssertThrowMPI`. I also rearranged things so that this exception and macro are only present when `DEAL_II_WITH_MPI` is true.

drwells commented:
One problem with the test I added is that mpich and openmpi print different messages. Is there a way to specify multiple output files to cover this?

tjhei commented:
> Is there a way to specify multiple output files to cover this?

Yes: https://www.dealii.org/developer/developers/testsuite.html#variants

bangerth commented on source/base/exceptions.cc:
```diff
@@ -336,6 +343,58 @@ void ExceptionBase::generate_message () const
+    // class.
+    int error_class = 0;
+    int ierr = MPI_Error_class (error_code, &error_class);
+    error_name_known = error_name_known && (ierr == MPI_SUCCESS);
```
This condition is funny because you have set `error_name_known` to `true` above :-) I think you can just declare and initialize the variable here with `(ierr == MPI_SUCCESS)`.

bangerth commented:
This looks good to me, up to that one small comment I just had.

As for `AssertThrowMPI`: I think it's quite obvious what the macro does -- and the fact that the variable that it takes as argument is clearly (by name) an error code is definitely also helpful. So I think there is (i) little confusion any reader could experience, and (ii) it removes a whole bunch of pretty repetitive boiler plate code. So I'm still for it. 

Anyone else with an opinion on the issue?

drwells commented:
Well, I am ambivalent about the macro, so I will wait to see if someone can break the tie :)

I cleaned up the history and fixed the issue with the boolean.

bangerth commented:
We still need a tie breaker!

tjhei commented:
> We still need a tie breaker!

I am okay with it as is. No need to drag out this war. ;-)

bangerth commented:
Since nobody objected, and since @tjhei waffled and showed how weak he is, I'm going to be the strong one and JUST HIT THAT BUTTON!

drwells commented:
ouch...

bangerth commented:
I suspect that @tjhei knows that I write from a perspective of friendship, love, and respect (as well as humor) between two people who have known each other for years and shared many miles on backroads ;-)
